src/gui/painting/qpainter.cpp

```diff
@@ -2673,6 +2673,152 @@ QPainterPath QPainter::clipPath() const
+{
+    QPainterPath tempPath;
+    tempPath.addRect(info.rectf);
+    tempPath = tempPath * matrix;
+    return tempPath;
+}
+
+static QPainterPath RectFClipIntersectClip(const QTransform &matrix, const QPainterClipInfo &info)
+{
+    // Use rect intersection if possible.
+    QPainterPath tempPath;
+
+    if (matrix.type() <= QTransform::TxScale) {
+        tempPath.addRect(matrix.mapRect(info.rectf));
+    } else {
+        tempPath.addRegion(matrix.map(QRegion(info.rectf.toRect())));
+    }
+    return tempPath;
+}
+
+static QPainterPath RegionClipReplaceClip(const QTransform &matrix, const QPainterClipInfo &info)
+{
+    QPainterPath tempPath;
+    tempPath.addRegion(info.region * matrix);
+    return tempPath;
+}
+
+static QPainterPath RegionClipIntersectClip(const QTransform &matrix, const QPainterClipInfo &info)
+{
+    return RegionClipReplaceClip(matrix, info);
+}
+
+enum ActionForClipOperation { 
+    NoClip, 
+    ReplaceClip, 
+    IntersectClip 
+};
+
+typedef QPainterPath (*PainterFunc)(const QTransform &matrix, const QPainterClipInfo &info);
+
+PainterFunc clipped_path[][IntersectClip + 1] = {
+    { noClipping, RegionClipReplaceClip, RegionClipIntersectClip },
+    { noClipping, PathClipReplaceClip, PathClipIntersectClip },
+    { noClipping, RectClipReplaceClip, RectClipIntersectClip },
+    { noClipping, RectFClipReplaceClip, RectFClipIntersectClip }
+};
+
+
+/*!
+    Returns the current clip path in logical coordinates, unrounded.
+
+    The main purpose for this function is inside QSvgGenerator,
+    where we do not want any points to be rounded during the
+    creation of the file.
+
+    \warning QPainter does not store the combined clip explicitly as
+    this is handled by the underlying QPaintEngine, so the path is
+    recreated on demandand transformed to the current logical
+    coordinate system.This is potentially an expensive operation.
+
+    \sa setClipPath(), clipRegion(), setClipping()
+*/
+QPainterPath QPainter::clipPathF() const
+{
+    Q_D(const QPainter);
+
+    if (!d->engine) {
+        qWarning("QPainter::clipPath: Painter not active");
+        return QPainterPath();
+    }
+
+    // No clip, return empty
+    if (d->state->clipInfo.isEmpty()) {
+        return QPainterPath();
+    }
+
+    // Update inverse matrix, used below.
+    if (!d->txinv)
+        const_cast<QPainter *>(this)->d_ptr->updateInvMatrix();
+
+    /*
+    * Modified version of "clipRegion" that is used in ::clipPath
+    * as "return qt_regionToPath(clipRegion());" that prevents rounding of values,
+    * placing them directly into a QPainterPath.
+    */
+    QPainterPath path;
+    bool initializing = true;
+
+    // ### Falcon: Use QPainterPath
+    for (const QPainterClipInfo &info : qAsConst(d->state->clipInfo)) {
+        QTransform matrix = (info.matrix * d->invMatrix);
+
+        if (initializing) {
+            path = clipped_path[info.clipType][ReplaceClip](matrix, info);
```

Comment:
In theory this can give the wrong result. It is perfectly valid (though nonsensical) to call QPainter::setClipPath with a valid QPainterPath and the Qt::NoClip operation. That would then mean that you add that path to the clipping when it was set with no clip and should be ignored.

I'm not sure I see the need for this separate handling of the first case. Does `path &= clipped_path[info.clipType][IntersectClip](matrix, info);` give weird results if the path is not initialised? That's the only way it could go wrong that I could see, so it might be better to put this handling there. Or you could check whether the operation == NoClip in this block.

Reply:
I guess intersect with an empty path would give an empty path, so it would need some kind of handling...